### PR TITLE
Add double backward support for CTC loss

### LINKED ISSUE #70108: Derivative for _ctc_loss_backward
### 🚀 The feature, motivation and pitch

Discussion motivation: https://discuss.pytorch.org/t/higher-order-gradients-of-ctcloss/35019

Hi! I'm trying to do double backward for CTC loss function and I can't do it due to `Runtime Error: derivative for _ctc_loss_backward is not implemented` error.

I'm implementing https://arxiv.org/abs/2102.08098 paper for a network with CTC loss. So I can't do it in a neutral way with default `torch.nn.functional.ctc_loss`.

### Alternatives

As an alternatives, I'm considering using different losses or hand-written ctc loss realization which are not optimal for our network, so it can cause not optimal optimization process.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Fierynx**
# Fixing an Issue

## Issue

Fixes #70108

## Summary

Adds double backward support for CTC loss by recomputing the CTC forward-backward variables with differentiable ATen operations when higher-order gradients are enabled. First-order backward continues to use the existing native, cuDNN, and MIOpen implementations.

## Testing

- `python test/test_nn.py -k ctc_loss` : 12 passed, 5 skipped
- CPU and CUDA gradgradcheck, including both list and tensor length inputs and `zero_infinity`: passed
- cuDNN CTC Hessian-vector product comparison against native CTC: passed
- Autograd code generation: passed
- `python -m spin quicklint`: passed
- `git diff --check`: passed

GPU validation was run on an NVIDIA GeForce RTX 3090 with CUDA and cuDNN.

## Checklist

- [x] Passes lint (`spin quicklint`)
- [x] Added/updated tests
- [ ] Updated documentation (not applicable)
- [ ] Included benchmark results (not applicable)

## BC-breaking?

No.

## AI assistance

I used Codex to assist with investigation and validation, as well as parts of the implementation in `torch/csrc/autograd/FunctionsManual.cpp`. I reviewed the final diff and test results.

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2751,10 +2751,10 @@
 
 # cudnn
 - name: _cudnn_ctc_loss(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank, bool deterministic, bool zero_infinity) -> (Tensor, Tensor)
-  log_probs: _cudnn_ctc_loss_backward(grad, result0, result1, zero_infinity)
+  log_probs: "GradMode::is_enabled() ? ctc_loss_differentiable_backward(grad, log_probs, targets, input_lengths, target_lengths, blank, zero_infinity) : _cudnn_ctc_loss_backward(grad, result0, result1, zero_infinity)"
```
Comment (Skylion007) [inline tools/autograd/derivatives.yaml]: You should just had a double backward derivative to the fused op derivative no reason to change impl based on is GradMoxd

**Review by Skylion007 — CHANGES_REQUESTED** [review]

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (Fierynx) [reply tools/autograd/derivatives.yaml]: Updated. First-order backward is unchanged, and double backward is now defined explicitly for the native and fused CTC ops.
